src/Language/PureScript/Interactive.hs

```diff
@@ -168,7 +169,10 @@ handleDecls
   :: (MonadReader PSCiConfig m, MonadState PSCiState m, MonadIO m)
   => [P.Declaration]
   -> m ()
-handleDecls ds = do
+handleDecls ds
+  | isJust (badDeclarationErrors ds) =
```

Comment:
`isJust` is a code smell: better to match on the result and do whatever it is you want to do with the `Just` value, I think.

Putting that aside for a moment though, did you consider putting this check in the parser instead? I think that would be a bit tidier.

Reply:
I could replace it with something like this instead:

```haskell
handleDecls'
  :: (MonadReader PSCiConfig m, MonadState PSCiState m, MonadIO m)
  => [P.Declaration]
  -> m ()
handleDecls' ds =
  maybe noBadDeclarations (printErrors . mconcat) (badDeclarationErrors ds) where
    noBadDeclarations = do
      st <- gets (updateLets (++ ds))
      let m = createTemporaryModule False st (P.Literal P.nullSourceSpan (P.ObjectLiteral []))
      e <- liftIO . runMake $ rebuild (map snd (psciLoadedExterns st)) m
      either printErrors (const (put st)) e
```

As for the parser bit, I will look into putting it there instead.

Reply:
It's a little bit awkward because we currently have this function that determines if a declaration is good:

```haskell
acceptable :: P.Declaration -> Bool
acceptable P.DataDeclaration{} = True
acceptable P.TypeSynonymDeclaration{} = True
acceptable P.ExternDeclaration{} = True
acceptable P.ExternDataDeclaration{} = True
acceptable P.TypeClassDeclaration{} = True
acceptable P.TypeInstanceDeclaration{} = True
acceptable P.ExternKindDeclaration{} = True
acceptable P.TypeDeclaration{} = True
-- acceptable (P.ValueDeclaration (P.ValueDeclarationData _ (P.Ident "it") _ _ _)) = False
acceptable P.ValueDeclaration{} = True
acceptable _ = False
```
I added the commented out line just to see how this would behave and with it you get the following behavior in the REPL:

```
> it = 4
(line 1, column 4):
unexpected =
expecting ., {, _, number, char literal, string literal, "true", "false", [, "\\", qualifier, data constructor name, identifier, "case", "if", "do", "ado", "let", (, hole literal, ::, operator, "import", declaration, expression or end of input
```

As opposed to the current behavior with a bit more explanation:

```
> it = 1
Error found:

  it is a variable used to refer to the previously evaluated expression in the REPL. Redefining it is not allowed.


See https://github.com/purescript/documentation/blob/master/errors/NoItInREPL.md for more information,
or to contribute content related to this error.
```

Is there possibly some point after this `acceptable` call where a declaration for `it` could be pre-empted (before the check in the declarations that I added, that is).

Reply:
Ah right yes, the error you get by modifying `acceptable` is quite poor. That must be happening because of our use of `try`. How about:

* pushing the use of `try` in the case of `parseDeclaration` slightly further down, so that it is outside of the `acceptable` check,
* modifying the `acceptable` check so that it returns an error where necessary rather than just a Bool?